modules/axtask/src/run_queue.rs

```diff
@@ -1,45 +1,216 @@
 use alloc::collections::VecDeque;
 use alloc::sync::Arc;
-use kspin::SpinNoIrq;
+use core::mem::MaybeUninit;
+use core::sync::atomic::{AtomicUsize, Ordering};
+
+use cpumask::CpuMask;
+use kernel_guard::BaseGuard;
+use kspin::SpinRaw;
 use lazyinit::LazyInit;
 use scheduler::BaseScheduler;
 
+use axhal::cpu::this_cpu_id;
+
 use crate::task::{CurrentTask, TaskState};
-use crate::{AxTaskRef, Scheduler, TaskInner, WaitQueue};
+use crate::{AxTaskRef, CpuSet, Scheduler, TaskInner, WaitQueue};
+
+macro_rules! percpu_static {
+    ($($name:ident: $ty:ty = $init:expr),* $(,)?) => {
+        $(
+            #[percpu::def_percpu]
+            static $name: $ty = $init;
+        )*
+    };
+}
+
+percpu_static! {
+    RUN_QUEUE: LazyInit<AxRunQueue> = LazyInit::new(),
+    EXITED_TASKS: VecDeque<AxTaskRef> = VecDeque::new(),
+    WAIT_FOR_EXIT: WaitQueue = WaitQueue::new(),
+    IDLE_TASK: LazyInit<AxTaskRef> = LazyInit::new(),
+}
+
+/// An array of references to run queues, one for each CPU, indexed by cpu_id.
+///
+/// This static variable holds references to the run queues for each CPU in the system.
+///
+/// # Safety
+///
+/// Access to this variable is marked as `unsafe` because it contains `MaybeUninit` references,
+/// which require careful handling to avoid undefined behavior. The array should be fully
+/// initialized before being accessed to ensure safe usage.
+static mut RUN_QUEUES: [MaybeUninit<&'static mut AxRunQueue>; axconfig::SMP] =
+    [ARRAY_REPEAT_VALUE; axconfig::SMP];
+const ARRAY_REPEAT_VALUE: MaybeUninit<&'static mut AxRunQueue> = MaybeUninit::uninit();
 
-// TODO: per-CPU
-pub(crate) static RUN_QUEUE: LazyInit<SpinNoIrq<AxRunQueue>> = LazyInit::new();
+/// Returns a reference to the current run queue.
+///
+/// ## Safety
+///
+/// This function returns a static reference to the current run queue, which
+/// is inherently unsafe. It assumes that the `RUN_QUEUE` has been properly
+/// initialized and is not accessed concurrently in a way that could cause
+/// data races or undefined behavior.
+///
+/// ## Returns
+///
+/// A static reference to the current run queue.
+// #[inline(always)]
+pub(crate) fn current_run_queue<G: BaseGuard>() -> AxRunQueueRef<'static, G> {
```

Comment:
Tried, but:
```bash
error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
  --> /home/hky/workspace/arceos-org/arceos/modules/axtask/src/run_queue.rs:58:33
   |
58 | pub(crate) fn current_run_queue<G: BaseGuard = NoOp>() -> AxRunQueueRef<'static, G> {
   |                                 ^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
   = note: `#[deny(invalid_type_param_default)]` on by default
```